# Lab 7: Predicting Voter Turnout

---

Welcome to Lab 7! This week, we will be constructing models to predict voter turnout. We will be using the same dataset as last week's lab (note that I removed `treat`).

# Before You Begin: Disable AI Assistance

To ensure you learn the concepts and complete this assignment based on your own understanding, you are *strongly* encouraged to turn off Google Colab's built-in generative AI features before you begin. This will also help you prepare for the midterms and final project.

**Follow these steps:**
1.  Go to the **Edit** menu at the top of the page.
2.  Click on **Notebook settings**.
3.  Check the box next to **"Hide generative AI features"**.
5.  Click **Save**.

This will prevent Google Gemini from suggesting or writing code for you, allowing you to focus on solving the problems yourself.




# Set-up

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
data = pd.read_csv("https://raw.githubusercontent.com/joshuakalla/data_science_campaigns/master/Colab/Lab7/gerber_huber_2014_data.csv")
data.head()

id  voted06  voted08  voted09  voted10  voted11  voted12  voted13  \
0  1989703        0        0        0        0        0        0        0   
1   555323        0        0        0        0        0        0        0   
2   915202        1        1        0        1        0        0        0   
3   839095        0        1        0        0        0        0        0   
4   197647        0        0        0        0        0        0        0   

   voted14   age  race_afam  race_hispanic  race_other  race_white  female  
0        0  26.0          0              0           0           1       1  
1        0  37.0          0              0           0           1       1  
2        1  26.0          1              0           0           0       0  
3        1  46.0          0              0           0           1       1  
4        0  21.0          0              0           0           1       0

# 1. Model-based predictions





Logistic Regression is a classification algorithm that is used to predict the probability of a categorical dependent variable. In logistic regression, the dependent variable is a binary variable that contains data coded as 1 (voted, yes, success, etc.) or 0 (did not vote, no, failure, etc.). In other words, the logistic regression model predicts P(Y=1) (probability that a person voted) as a function of X (many different covariates).


(You could use other algorithms to build a model. We are using logistic regression because it is fast and easy.)

Our goal is to predict whether an individual voter *i* voted in the 2014 election as a function of other features we know about them (age, past vote history, race, and gender). With a model of voting in the 2014 election, a campaign in the future (such as in 2018) could better target voters.

To get started, let's see if there are any big differences in who votes in 2014. Run the below code.

In [2]:
data.groupby('voted14').mean()

id   voted06   voted08  voted09   voted10  voted11  \
voted14                                                                 
0        1.202580e+06  0.227614  0.409396      0.0  0.161182      0.0   
1        1.214615e+06  0.396483  0.661373      0.0  0.470221      0.0   

          voted12  voted13        age  race_afam  race_hispanic  race_other  \
voted14                                                                       
0        0.216056      0.0  38.491964   0.049669       0.027084    0.100088   
1        0.723908      0.0  45.649620   0.043250       0.017867    0.143222   

         race_white    female  
voted14                        
0          0.823158  0.698015  
1          0.795661  0.766024

**Question 1.** What do you find? Do any individual variables seem to be more or less predictive of voting in 2014? Interpret the above table.

**Answer the question here.**

We are now going to build a simple model. We first need to define our outcome variable (did someone vote in 2014 or not, call this *Y*) and our set of predictor variables (call this a matrix *X*). In this first simple case, we will only include `voted12` and `age` in X.

In [3]:
cols = ['voted12', 'age', 'intercept']
data['intercept'] = 1 # add a column of 1's for the intercept term
X = data[cols]
X.head()

voted12   age  intercept
0        0  26.0          1
1        0  37.0          1
2        0  26.0          1
3        0  46.0          1
4        0  21.0          1

In [4]:
y = data['voted14']
y.head()

0    0
1    0
2    1
3    1
4    0
Name: voted14, dtype: int64

We will now build our model. Note that we are now using a new module in Python called `statsmodels`. You can learn more about this module [here](https://www.statsmodels.org/stable/index.html).

In [5]:
import statsmodels.api as sm
## ignore the warning; nothing to worry about
logit_model=sm.Logit(y,X)
result=logit_model.fit()
print(result.summary2())

Optimization terminated successfully.
         Current function value: 0.442076
         Iterations 6
                         Results: Logit
Model:              Logit            Method:           MLE       
Dependent Variable: voted14          Pseudo R-squared: 0.193     
Date:               2025-11-05 20:55 AIC:              26284.7836
No. Observations:   29722            BIC:              26309.6825
Df Model:           2                Log-Likelihood:   -13139.   
Df Residuals:       29719            LL-Null:          -16285.   
Converged:          1.0000           LLR p-value:      0.0000    
No. Iterations:     6.0000           Scale:            1.0000    
------------------------------------------------------------------
              Coef.   Std.Err.     z      P>|z|    [0.025   0.975]
------------------------------------------------------------------
voted12       2.1814    0.0314   69.4384  0.0000   2.1198   2.2430
age           0.0145    0.0008   17.2811  0.0000   0.0129   0.

**Question 2.** Can you interpret this output? What do you think this means? (You might want to give [this](https://www.juanshishido.com/logisticcoefficients.html) a read.)

**Answer the question here.**

With the below code, we can construct model scores. Remember that the output of this can be understood as the probability that someone votes in 2014 as a function of their age and whether they voted in 2012.

In [6]:
y_pred = result.predict(X)
y_pred.head()

0    0.082433
1    0.095372
2    0.082433
3    0.107281
4    0.077096
dtype: float64

**Question 3.** Make a plot showing the relationship between your predicted model (`y_pred`) and whether someone actually voted in 2014. In text, make sure you interpret this plot.

*Hint.* You will probably want to use the [Pandas.cut() method](https://www.geeksforgeeks.org/pandas-cut-method-in-python/). This will allow you to create bins of `y_pred`. Remember the figures we discussed in the Likely Voter slides. You probably want to make a similar plot to this, like the below figure. (You don't need to exactly copy this, just an example of what you could do.) **This is a challenging question.**

**This example shows more or less how the final plots should look.**

![](https://raw.githubusercontent.com/joshuakalla/data_science_campaigns/master/Colab/Lab7/sample_plot.png)


In [7]:
# insert plot and interpretation

# 2. Confusion matrix and model metrics

To more formally assess this model, let's make a 2x2 confusion matrix, like we did in class. The confusion matrix needs to have 4 cells: number of true positives, number of true negatives, number of false positives, number of false negatives. (If you want a reminder on a confusion matrix, see the slides from lecture or read [this](https://www.python-course.eu/confusion_matrix.php).)

For this exercise, we are going to define our threshold as 0.5. That means that if someone's predicted turnout (`y_pred`) is greater than 0.5, we are going to say that we predicted they vote. If their score is less than or equal to 0.5, we are going to say we predicted they did not vote. (Note that this threshold is somewhat arbitrary. We could define different cut-offs.)

**Question 4.** Make the confusion matrix. Calculate the number of true positives, number of true negatives, number of false positives, number of false negatives. Fill in the table.

In [8]:
# insert your code here

Fill in the below table with the correct numbers.

- Actual Negative and Predicted Negative (what do we call this?):
- Actual Negative and Predicted Positive (what do we call this?):
- Actual Positive and Predicted Negative (what do we call this?):
- Actual Positive and Predicted Positive (what do we call this?):

**Question 5.** Based on this table, calculate the model's accuracy, precision, and recall. Interpret this.

In [9]:
# insert your code here

**Put your interpretation here.**

**Question 6.** Can a different model do better? It is now your turn to build a model from scratch. Follow the same steps as above. Select your predictor variables. Build your model. Construct a confusion matrix. Calculate accuracy, precision, and recall. How does this model do? Does it do better or worse than the original model? Would you use it?

In [10]:
# insert your code here

**Put your interpretation here.**

**Question 7.** Find a way to plot the differences between the two models we built.

In [11]:
# insert your code here

# 3. Are we overfitting?

If you don't remember what overfitting means, review your notes from class or give this a read: https://www.ibm.com/topics/overfitting

Let's see how these models do out-of-sample. Are we overfitting?

To test this, you will need to run your model on a new data set (test set). You can then assess the confusion matrix and accuracy/precision/recall of your model on that test set.

Note that a logistic regression is written as:

$p(x) = \frac{1}{1+e^{-(\beta_0 + \beta_1 x)}}$, where $e$ is Euler's number, approximately 2.71828. You can get Euler's number in Python using `np.exp()`.

We will need to manually re-create this logistic regression. From my model, the coefficients are:

- voted12: 2.1814
- age: 0.0145
- intercept: -2.7879

We could therefore recreate our logistic regression by calculating:

$p(voted14) = \frac{1}{1+e^{-(-2.7879 + 0.0145*age + 2.1814*voted12)}}$

Let's see this in action.

In [12]:
# Calculate y_pred by hand; call it y_pred2
y_pred2 = 1/(1+np.exp(-(-2.787942 + 2.181396 *X["voted12"] + .0145461 * X["age"])))

# Confirm that y_pred is the same as y_pred2
y_preds = pd.concat([pd.DataFrame(y_pred), pd.DataFrame(y_pred2)], axis = 1)
y_preds.columns = ['y_pred', 'y_pred2']
y_preds['diff'] = abs(y_preds['y_pred']) - y_preds['y_pred2']

# Print out this data frame
print(y_preds.head())
print(y_preds["diff"].mean()) # Small differences due to rounding, but this is essentially a 0
# These columns are the same

     y_pred   y_pred2          diff
0  0.082433  0.082433 -7.398153e-08
1  0.095372  0.095372 -1.066650e-07
2  0.082433  0.082433 -7.398153e-08
3  0.107281  0.107281 -1.386375e-07
4  0.077096  0.077096 -6.124326e-08
-1.8630459050923657e-07


I now want you to test if your model is over-fit. Don't cheat by recreating your model. Just use what you have above. I care about the process, not the actual performance.

The data you used above came from South Dakota. I want you to see how the model performs in Wisconsin.

In [13]:
test = pd.read_csv("https://raw.githubusercontent.com/joshuakalla/data_science_campaigns/master/Colab/Lab7/test_data.csv")
test.head()

id  voted06  voted08  voted09  voted10  voted11  voted12  voted13  \
0   551662        0        0        0        0        0        1        0   
1  1057779        1        1        0        1        0        1        0   
2   248929        1        1        0        0        0        1        0   
3   723382        1        1        0        1        0        1        0   
4  2259556        1        0        0        0        1        1        0   

   voted14   age  race_afam  race_hispanic  race_other  race_white  female  
0        1  75.0          0              0           0           1       1  
1        0  88.0          0              0           0           1       1  
2        0  28.0          0              0           0           1       0  
3        1  50.0          0              0           0           1       1  
4        1  55.0          0              0           0           1       1

**Question 8.** Evaluate how your model performs out-of-sample using `test_data.csv` and the coefficients from your model. Create a confusion matrix and calculate the accuracy/precision/recall. How does your model perform out-of-sample?

# Congratulations!

You are done with the lab. Before you finish and submit, please fill out this brief evaluation:

- I spent around XXXX hours on this lab,.
- This lab was (too easy, too hard, just about the right difficulty).

All assignments in the course will be distributed as notebooks like this one, and you will submit your work as a PDF.  

# How to Convert your Colab notebook to a PDF and download it

Follow these instructions exactly to make sure your notebook is correctly converted to a PDF and saved to your computer.

---

## 1. Check everything is in order and all the code runs

Before starting the conversion process, make sure your notebook is complete and error-free.

1. **Open your notebook** in Google Colab.
2. **Save your work**:
   - Go to **File → Save** or press `Ctrl + S` (Windows) / `Cmd + S` (Mac).
   - This ensures that all your recent changes are stored.
3. **Run all the cells** to confirm there are no errors:
   - In the top menu, select **Cell → Run All**.
   - Colab will execute every cell in order.
4. **Watch for errors**:
   - If you see any red error messages, fix them before proceeding.
   - A notebook with errors will **not** convert to PDF correctly.
5. Once all cells run without errors, you can proceed to the next step.

---

## 2. Make sure your notebook is saved in Google Drive and named correctly
Before converting, confirm that your notebook is stored in your Google Drive inside the `Colab Notebooks` folder.

1. Look at the top-left corner of the Colab page — you’ll see the notebook’s current name next to two yellow circle icons.
2. Rename the file by directly clicking on the name, type the new name, and press **Enter**.  
  - Please rename the notebook to LASTNAME_FIRSTNAME_LAB#.pdf. So for this lab, I would call it Alberto_Stefanelli_Lab7.ipynb.
  - The name must end with `.ipynb`
3. Ensure the notebook is in your Google Drive (not in Colab’s temporary session storage):  
   - In the menu, click **File → Locate in Drive**.  
   - This will open the folder in Drive where the notebook is stored.  
   - If it’s not in `My Drive/Colab Notebooks`, move it there for easier access.

---

## 3. Install the required tools

We wrote some code (see below) to automatically convert your Notebook to PDF. When you run the provided code cell, the first step will install some essential pieces of software (i.e., Pandoc and Latex) inside your Colab environment. There is no need to exactly understand what is happening

**Important:**
- This installation will take between 2 to 5 minutes.
- Do **not** close or refresh the Colab page while it runs.

---

## 4. Mount your Google Drive in Colab

After installing the requirements, the code will ask Colab to **mount** your Google Drive. You can use both your personal or Yale account. This is needed because the notebook you are converting must be saved in Drive before it can be converted to PDF.

You will see a pop-up with a **link**:
1. Click the link.
2. Sign in to your Google account (use the same account where your Colab notebook is saved).
3. If prompted, copy the long **authorization code** provided.
  - Paste that code into the input box in Colab and press **Enter**.
5. This will connect your Google Drive to Colab
6. If the link does not appear, make sure your browser is not blocking pop-ups.

---

## 5. Enter your notebook’s file name

The code will now ask to enter your notebook’s exact file name

1. Type the **full name** of your notebook, including the `.ipynb` ending.  
  - Example: `Alberto_Stefanelli_Lab7.ipynb.`
2. Make sure the name matches exactly, including capitalization and underscores.
3. Press **Enter**.

---

## 7. Convert the notebook to PDF and download it

The code will convert the notebook file to a pdf. After the PDF is created, your browser will show a download pop-up or automatically save the file to your Downloads folder. You can now open the PDF with any PDF reader. Once you have your PDF and made sure eveything is in order, you can then upload it to Canvas.

---

## 8. If you see an error:

Double-check that:

- All cells run without errors
- Mounted Google Drive without errors
- Saved notebook to Google Drive and not locally or on Github
- Entered correct notebook file name with `.ipynb`.
- Conversion completed without any errors.
- PDF downloaded to your computer (download and pop-ups are not blocked by your browser)


**Fallback:** If the PDF export method below fails (for example, due to LaTeX or pandoc errors), you can use https://convert.ploomber.io/ as a fallback option. However, I strongly suggest trying the methods below first and using this fallback only as a last resort.

**If you run into any issues, please reach out for help**

In [14]:
from google.colab import drive, files

drive.mount('/content/drive', force_remount=True)

# Ask for the notebook name
notebook_name = input(
    "Enter your notebook’s exact file name,\n"
    "exactly as shown in the top-left corner of the Colab page (next to the two yellow circle icons): "
)

# Build paths
input_path = f"/content/drive/MyDrive/Colab Notebooks/{notebook_name}"
output_path = input_path.replace(".ipynb", ".pdf")

# ---- Convert using WebPDF (handles remote images/URLs) ----
!pip -q install "nbconvert[webpdf]>=7" "pyppeteer>=2"

!jupyter nbconvert \
  --to webpdf \
  --allow-chromium-download \
  "{input_path}"

files.download(output_path)

Mounted at /content/drive
Enter your notebook’s exact file name,
exactly as shown in the top-left corner of the Colab page (next to the two yellow circle icons): lab07.ipynb
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.1/50.1 kB 2.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.9/84.9 kB 4.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.9/82.9 kB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 144.2/144.2 kB 7.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.9/37.9 MB 21.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 625.7/625.7 kB 20.0 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
dataproc-spark-connect 0.8.3 requires websockets>=14.0, but you have websockets 10.4 which is incompatible.
google-adk 1.17.0 requir

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>